tmt/result.py

```diff
@@ -317,11 +334,18 @@ def from_test_invocation(
             ids=ids,
             log=log or [],
             guest=ResultGuestData.from_test_invocation(invocation=invocation),
-            data_path=invocation.relative_test_data_path)
+            data_path=invocation.relative_test_data_path,
+            check=invocation.check_results or [])
 
-        return _result.interpret_result(invocation.test.result)
+        interpret_checks = {check.how: check.result for check in invocation.test.check}
```

Comment:
This way doesn't really work when multiple check.how/check.name are the same. There surely has to be a better way to add/match `CheckResultInterpret` to each `Check`?

Reply:
I believe we spoke about a way: there is a trivial mapping between a result and its parent test, `TestInvocation` stores results & points to the test. We did not needed to transition from check result to a check, therefore there is no such mapping yet, we need to establish one. My proposal was to add a dedicated mapping to `TestInvocation` class which would map from `CheckResult` to `Check` instances, and it would be populated in `_run_checks_for_test()` (most likely), as that is the place where check results are born. Something like this:

```diff
diff --git a/tmt/steps/execute/__init__.py b/tmt/steps/execute/__init__.py
index 407b4ac0..7f77d179 100644
--- a/tmt/steps/execute/__init__.py
+++ b/tmt/steps/execute/__init__.py
@@ -16,6 +16,7 @@ import fmf.utils
 
 import tmt
 import tmt.base
+import tmt.checks
 import tmt.log
 import tmt.steps
 import tmt.utils
@@ -175,6 +176,9 @@ class TestInvocation:
     results: list[Result] = dataclasses.field(default_factory=list)
     check_results: list[CheckResult] = dataclasses.field(default_factory=list)
 
+    check_result_to_check: dict[CheckResult, tmt.checks.Check] = dataclasses.field(
+        default_factory=dict)
+
     check_data: dict[str, Any] = field(default_factory=dict)
 
     return_code: Optional[int] = None
@@ -954,6 +958,8 @@ class ExecutePlugin(tmt.steps.Plugin[ExecuteStepDataT, None]):
                 result.end_time = format_timestamp(timer.end_time)
                 result.duration = format_duration(timer.duration)
 
+                invocation.check_result_to_check[result] = check
+
             results += check_results
 
         return results
```

It's not clean, it's not nice, I'd like to make it look and feel like the test invocation, but I don't have that ready and I promise to polish it later. With this, the following should work: use this mapping to collect check result/interpretation pairs, and use the right one for the given check result when calling the interpretation method:

```python
interpret_checks = {
    check_result: check.result for check_result, check in invocation.check_result_to_check.items()
}

...

check_result.interpret_check_result(interpret_checks[check_result]))
```

Reply:
oh, now I remembered I tried to go this way, but `CheckResult` is unhashable. And I can't really add `frozen=True` to it to make it so. Am I missing something @happz ?

Reply:
Ouch, I forgot about that, it's not immutable :/

Hm, how about going with what you have already, leaving the "multiple checks with the same `how`" case for later? Because if I'm not mistaken, this case is broken in more places than here. For example, checks fill `CheckResult.name` with their name, e.g. https://github.com/teemtee/tmt/blob/main/tmt/checks/dmesg.py#L199, therefore one check running twice will produce two results with the same attributes. So, more work to make this feature correctly supported.

For now, I'd say "running one check multiple times for the same test is not supported yet" and move on to finish the PR, and we would make the code right later.

Reply:
Sounds good.

Reply:
Seems to be covered.